### PR TITLE
Convert read stage tracking to new small_vector container

### PR DESCRIPTION
**Pull request by jzulauf-lunarg**
Add a "small vector optimized" vector work-alike container and convert
sync last_reads to use it instead of the fixed sized array.

small vector interface is a strict subset of std::vector, implementing only the entry-points needed, and in STL not google style s.t. that it functions with range based for algorithms, etc.

### CODE REVIEW

File: layers/vk_layer_data.h [hunk 1]
```diff
@@ -15,15 +15,223 @@
...
+            } else {
+                // Other is using the small_store
+                auto source = other.begin();
+                iterator dest;
+                if (large_store_) {
+                    // If this is using large store to a wholesale clobber of it.
+                    clear();
+                    dest = GetWorkingStore();
+                } else {
+                    // This is also using small store, so move assign were both have valid values
```
Comment (jeremyg-lunarg) [inline layers/vk_layer_data.h:128]: ... assign where both ...

**Review by jeremyg-lunarg — CHANGES_REQUESTED** [review]
Code LGTM, just some comment issues in the small_vector class

**Comment by jzulauf-lunarg** [comment]
@jeremyg-lunarg  -- revisions taken

File: layers/vk_layer_data.h [hunk 2]
```diff
@@ -15,15 +15,230 @@
...
+#include <limits>
+#include <memory>
 #include <unordered_map>
 #include <unordered_set>
 
+// A vector class with "small string optimization" -- meaning that the class contains a fixed working store for N elements.
+// Useful in in situations where the needed size is unknown, but the typical size is known  If size increases beyond the
+// fixed capacity, a dynamically allocated working store is created.
+//
+// NOTE: Unlike std::vector which only requires T to be CopyAssignable and CopyConstructable, small_vector requires T to be
```
Comment (jeremyg-lunarg) [inline layers/vk_layer_data.h:36]: thank you

**Review by jeremyg-lunarg — APPROVED** [review]
LGTM as long as CI still agrees.

**Pull request merged**